gamemodes/terrortown/gamemode/server/sv_roleselection.lua

```diff
@@ -0,0 +1,496 @@
+		if not roleCount then continue end
+
+		local curCount = 0
+		local amount = #forcedPlys
+
+		for i = 1, amount do
+			local pick = math.random(#forcedPlys)
+			local ply = forcedPlys[pick]
+
+			if roleCount <= curCount then break end -- if the limit is reached, stop selection for this role
+
+			table.remove(forcedPlys, pick) -- remove selected player to avoid multiple selection
+
+			roleselection.forcedRoles[tostring(ply:SteamID64())] = nil
+
+			if roleselection.finalRoles[ply] then continue end -- we don't need to set a final role if this player already has a final role
+
+			roleselection.finalRoles[ply] = subrole
+			curCount = curCount + 1
+
+			hook.Run("TTT2ReceivedForcedRole", ply, rd)
+
+			selectedPlys[ply] = true
+		end
+	end
+
+	roleselection.forcedRoles = {}
+
+	return selectedPlys
+end
+
+local function UpgradeRoles(plys, prevRoles, roleData, selectableRoles)
+	local availableRoles = {}
+
+	-- now upgrade this role if there are other subroles
+	for v in pairs(selectableRoles) do
+		if v.baserole == roleData.index then
+			availableRoles[#availableRoles + 1] = v
+		end
+	end
+
+	SetSubRoles(plys, prevRoles, availableRoles, selectableRoles)
+end
+
+---
+-- This function modifies the given `plys` var
+local function SelectBaseRolePlayers(plys, prevRoles, roleData, roleAmount)
+	local curRoles = 0
+	local plysList = {}
+
+	local minKarmaCVar = GetConVar("ttt_" .. roleData.name .. "_karma_min")
+	local min_karmas = minKarmaCVar and minKarmaCVar:GetInt() or 0
+
+	while curRoles < roleAmount and #plys > 0 do
+		-- select random index in plys table
+		local pick = math.random(#plys)
+
+		-- the player we consider
+		local ply = plys[pick]
+
+		-- give this player the role if
+		if roleData == INNOCENT -- this role is an innocent role
+			or #plys <= roleAmount -- or there aren't enough players anymore to have a greater role variety
+			or ply:GetBaseKarma() > min_karmas -- or the player has enough karma
+				and table.HasValue(prevRoles[ROLE_INNOCENT], ply) -- and the player was innocent before
+				and not ply:GetAvoidRole(roleData.index) -- and the player doesn't avoid this role
+			or math.random(3) == 2 -- or if the randomness decides
+		then
+			table.remove(plys, pick)
+
+			curRoles = curRoles + 1
+			plysList[curRoles] = ply
+
+			roleselection.finalRoles[ply] = roleData.index -- give the player the final baserole (maybe he will receive his subrole later)
+		end
+	end
+
+	return plysList
+end
+
+---
+-- Select selectable @{ROLE}s for a given list of @{Player}s
+-- @note This automatically synces with every connected @{Player}
+-- @param ?table plys list of @{Player}s. `nil` to calculate automatically
+-- @param ?number maxPlys amount of maximum @{Player}s. `nil` to calculate automatically
+-- @realm server
+function roleselection.SelectRoles(plys, maxPlys)
+	roleselection.selectableRoles = nil -- reset to enable recalculation
+
+	local prevRoles = {}
+
+	GAMEMODE.LastRole = GAMEMODE.LastRole or {}
```

Comment:
Why does this have to be stored on the GAMEMODE instead of the roleselection table?